### PR TITLE
DWR-1494 Add 'may-contain-embargoed-results' flag to aggregate report…

### PR DESCRIPTION
**Pull request by agorina**
… result - round 2

### CODE REVIEW

File: aggregate-service/src/main/java/org/opentestsystem/rdw/olap/repository/impl/JdbcCustomAggregateReportRowRepository.java [hunk 1]
```diff
@@ -125,15 +147,62 @@ private ReportRow removeFromTemplateIfExists(final Set<ReportRow> template, fina
...
+
         private ActiveAssessment findByIdAndSchoolYear(final String assessmentTypeCode, final int assessmentId, int schoolYear) {
             for (final ActiveAssessment activeAssessment : activeAssessmentRepository.findAllByTypeCode(assessmentTypeCode)) {
                 if (activeAssessment.getId() == assessmentId && activeAssessment.getExamSchoolYear() == schoolYear)
                     return activeAssessment;
             }
             throw new IllegalArgumentException("unable to find an assessment by id " + assessmentId + " and school year " + schoolYear);
         }
+
+        private Organization findOneByIdAndType(final long id, final OrganizationType type) {
```
Comment (agorina) [inline aggregate-service/src/main/java/org/opentestsystem/rdw/olap/repository/impl/JdbcCustomAggregateReportRowRepository.java]: This is unrelated but it came up during my first attempt to support embargo and I think it is a good change, simplifies the SQL for report.

File: aggregate-service/src/main/java/org/opentestsystem/rdw/olap/repository/impl/JdbcOrganizationRepository.java [hunk 2]
```diff
@@ -82,4 +87,41 @@
             throw ex;
         }
     }
+
+    /**
```
Comment (agorina) [inline aggregate-service/src/main/java/org/opentestsystem/rdw/olap/repository/impl/JdbcOrganizationRepository.java]: Unrelated, due to the SQL simplification.

File: aggregate-service/src/main/java/org/opentestsystem/rdw/olap/service/impl/DefaultCustomAggregateService.java [hunk 3]
```diff
@@ -35,16 +33,16 @@
     }
 
     @Override
-    public List<ReportRow> findByQuery(final User user, final ReportQuery query) {
+    public AggregateReport findByQuery(final User user, final ReportQuery query) {
         authorizationService.checkAuthorization(user, query);
-        checkIsValid(query);
+        checkIsValid(ensureDimension(query, Overall));
```
Comment (agorina) [inline aggregate-service/src/main/java/org/opentestsystem/rdw/olap/service/impl/DefaultCustomAggregateService.java]: Oops, I will do the bug fix here.

File: aggregate-service/src/main/resources/application.sql.yml [hunk 4]
```diff
@@ -43,6 +43,13 @@ sql:
   strictBoolean:
     findAll: select id, code from strict_boolean order by id
 
+  embargo:
+    isStateEmbargoed: >-
+      SELECT EXISTS(SELECT 1 FROM state_embargo WHERE aggregate = 1) as embargo_enabled
+
+    findAllEmbargoedDistrictIds: >-
+      SELECT id FROM district d WHERE EXiSTS (SELECT 1 FROM school WHERE embargo_enabled = 1 AND district_id = d.id)
```
Comment (agorina) [inline aggregate-service/src/main/resources/application.sql.yml]: @malaffoon , yes - this probably means we should have migrate districts, but...it only runs once a day.

File: aggregate-service/src/main/resources/custom-aggregate-report-template.sql.yml [hunk 5]
```diff
@@ -38,9 +38,7 @@ sqlbuilder:
               clauses:
                 select: >-
                      'State' AS organization_type,
-                     NULL AS organization_id,
-                     NULL AS organization_natural_id,
-                     '${app.state.name}' AS organization_name
+                     NULL AS organization_id
```
Comment (agorina) [inline aggregate-service/src/main/resources/custom-aggregate-report-template.sql.yml]: This is the SQL simplification mentioned above.

File: aggregate-service/src/test/java/org/opentestsystem/rdw/olap/repository/impl/JdbcCustomAggregateReportRowRepositoryEmbargoRST.java [hunk 6]
```diff
@@ -4,24 +4,23 @@
 import org.junit.Test;
```
Comment (agorina) [inline aggregate-service/src/test/java/org/opentestsystem/rdw/olap/repository/impl/JdbcCustomAggregateReportRowRepositoryEmbargoRST.java]: This is probably the most important file to review. I would really appreciate if somebody verifies my test cases here.

File: aggregate-service/src/test/java/org/opentestsystem/rdw/olap/sqlbuilder/CustomAggregateQueryProviderRST.java [hunk 7]
```diff
@@ -63,9 +63,10 @@
             "sum(CASE WHEN performance_level = 2 THEN 1 ELSE 0 END) AS level2, " +
             "sum(CASE WHEN performance_level = 3 THEN 1 ELSE 0 END) AS level3, " +
             "sum(CASE WHEN performance_level = 4 THEN 1 ELSE 0 END) AS level4, " +
```
Comment (agorina) [inline aggregate-service/src/test/java/org/opentestsystem/rdw/olap/sqlbuilder/CustomAggregateQueryProviderRST.java]: Once again - due to SQL change.

File: aggregate-service/src/main/java/org/opentestsystem/rdw/olap/repository/impl/JdbcCustomAggregateReportRowRepository.java [hunk 8]
```diff
@@ -125,15 +147,62 @@ private ReportRow removeFromTemplateIfExists(final Set<ReportRow> template, fina
...
+            }
+
+            //schools inherit their district embargo flag
+            final Long districtId = organization.getOrganizationType().equals(District) ? organization.getId() : ((School) organization).getDistrictId();
+            reportBuilder.isCreatedWhileDataEmbargoed(isDistrictEmbargoed(districtId));
+            return rowToVerify;
+        }
+
+        private boolean isDistrictEmbargoed(final long districtId) {
+            for (final Long embargoedDistrictIds : embargoRepository.findAllEmbargoedDistrictIds()) {
```
Comment (malaffoon) [inline aggregate-service/src/main/java/org/opentestsystem/rdw/olap/repository/impl/JdbcCustomAggregateReportRowRepository.java]: `return embargoRepository.findAllEmbargoedDistrictIds().contains(districtId);`

File: aggregate-service/src/main/java/org/opentestsystem/rdw/olap/repository/impl/JdbcEmbargoRepository.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    private String findAllEmbargoedDistrictIdsQuery;
+
+    JdbcEmbargoRepository(final NamedParameterJdbcTemplate template) {
+        this.template = template;
+    }
+
+    @Cacheable(value = AggregateServiceCache.Embargo, key = "#root.methodName")
+    @Override
+    public boolean isStateEmbargoed() {
+        return template.getJdbcOperations().queryForObject(isStateEmbargoedQuery, Boolean.class);
```
Comment (malaffoon) [inline aggregate-service/src/main/java/org/opentestsystem/rdw/olap/repository/impl/JdbcEmbargoRepository.java]: If you're only ever doing JdbcOperations why not just inject a JdbcTemplate instead?

File: aggregate-service/src/test/java/org/opentestsystem/rdw/olap/repository/impl/JdbcCustomAggregateReportRowRepositoryEmbargoRST.java (hunk 6 above)
Comment (malaffoon) [reply aggregate-service/src/test/java/org/opentestsystem/rdw/olap/repository/impl/JdbcCustomAggregateReportRowRepositoryEmbargoRST.java]: I'll have to download the patch to do that properly ... i'll try to get to it later today.

File: aggregate-service/src/test/java/org/opentestsystem/rdw/olap/repository/impl/JdbcCustomAggregateReportRowRepositoryEmbargoRST.java [hunk 10]
```diff
@@ -103,11 +111,13 @@ public void isShouldReturnSummativeExamsForStateWithEmbargoReleased() {
     })
     @Test
     public void isShouldReturnSummativeExamForStateWithEmbargoEnabledForEmbargoEdminOnly() {
```
Comment (malaffoon) [inline aggregate-service/src/test/java/org/opentestsystem/rdw/olap/repository/impl/JdbcCustomAggregateReportRowRepositoryEmbargoRST.java]: nit: `...AdminOnly()`

File: common/src/main/java/org/opentestsystem/rdw/reporting/common/model/AggregateReport.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
+package org.opentestsystem.rdw.reporting.common.model;
+
+import java.util.Collection;
+import java.util.List;
+
+import static com.google.common.collect.Lists.newArrayList;
+
+public class AggregateReport {
+    private List<ReportRow> rows;
+    private boolean isCreatedWhileDataEmbargoed;
```
Comment (malaffoon) [inline common/src/main/java/org/opentestsystem/rdw/reporting/common/model/AggregateReport.java]: Variable should be just `createdWhileDataEmbargoed`, then the getter will be `isCreatedWhileDataEmbargoed`.

File: common/src/main/java/org/opentestsystem/rdw/reporting/common/model/AggregateReport.java [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+            return this;
+        }
+
+        public Builder row(final ReportRow rows) {
+            this.rows.add(rows);
+            return this;
+        }
+
+        public Builder isCreatedWhileDataEmbargoed(final boolean isCreatedWhileDataEmbargoed) {
+            this.isCreatedWhileDataEmbargoed = isCreatedWhileDataEmbargoed;
```
Comment (malaffoon) [inline common/src/main/java/org/opentestsystem/rdw/reporting/common/model/AggregateReport.java]: ditto

File: common/src/main/java/org/opentestsystem/rdw/reporting/common/model/AggregateResponseMessage.java [hunk 13]
```diff
@@ -8,6 +8,7 @@
     private long id;
     private ReportStatus status;
     private long resultCount;
+    private boolean isCreatedWhileDataEmbargoed;
```
Comment (malaffoon) [inline common/src/main/java/org/opentestsystem/rdw/reporting/common/model/AggregateResponseMessage.java]: ditto (okay, i'll stop now)

File: aggregate-service/src/main/java/org/opentestsystem/rdw/olap/repository/impl/JdbcEmbargoRepository.java (hunk 9 above)
Comment (agorina) [reply aggregate-service/src/main/java/org/opentestsystem/rdw/olap/repository/impl/JdbcEmbargoRepository.java]: I do not know...I did not care..I was tired. Will change, if we decide to merge it.

File: common/src/main/java/org/opentestsystem/rdw/reporting/common/model/AggregateReport.java (hunk 11 above)
Comment (agorina) [reply common/src/main/java/org/opentestsystem/rdw/reporting/common/model/AggregateReport.java]: Oh yeah...it should...

File: aggregate-service/src/test/java/org/opentestsystem/rdw/olap/repository/impl/JdbcCustomAggregateReportRowRepositoryEmbargoRST.java (hunk 6 above)
Comment (agorina) [reply aggregate-service/src/test/java/org/opentestsystem/rdw/olap/repository/impl/JdbcCustomAggregateReportRowRepositoryEmbargoRST.java]: Thank you, this is very much appreciated.

File: common/src/main/java/org/opentestsystem/rdw/reporting/common/model/AggregateResponseMessage.java (hunk 13 above)
Comment (agorina) [reply common/src/main/java/org/opentestsystem/rdw/reporting/common/model/AggregateResponseMessage.java]: Why stop now? This looks like fun, you should keep going :)

File: aggregate-service/src/test/java/org/opentestsystem/rdw/olap/repository/impl/JdbcCustomAggregateReportRowRepositoryEmbargoRST.java (hunk 6 above)
Comment (malaffoon) [reply aggregate-service/src/test/java/org/opentestsystem/rdw/olap/repository/impl/JdbcCustomAggregateReportRowRepositoryEmbargoRST.java]: I pulled down the patch and i think it is pretty good coverage.

Comment (agorina) [reply aggregate-service/src/test/java/org/opentestsystem/rdw/olap/repository/impl/JdbcCustomAggregateReportRowRepositoryEmbargoRST.java]: Thank you.

File: common/src/main/java/org/opentestsystem/rdw/reporting/common/model/AggregateResponseMessage.java [hunk 14]
```diff
@@ -17,6 +18,13 @@ public long getId() {
         return id;
     }
 
+    /**
+     * @return true if the result may contain embargo enabled data, false otherwise
+     */
+    public boolean isCreatedWhileDataEmbargoed() {
+        return createdWhileDataEmbargoed;
+    }
```
Comment (wtritch) [inline common/src/main/java/org/opentestsystem/rdw/reporting/common/model/AggregateResponseMessage.java]: So this is the flag that determines whether or not the report can be downloaded and a warning is displayed to the user that "The results may contain embargoed results and may not be downloaded" ?

File: aggregate-service/src/main/java/org/opentestsystem/rdw/olap/repository/impl/JdbcCustomAggregateReportRowRepository.java [hunk 15]
```diff
@@ -125,15 +147,59 @@ private ReportRow removeFromTemplateIfExists(final Set<ReportRow> template, fina
...
          */
+        private ReportRow processEmbargo(final ReportRow rowToVerify, final AggregateReport.Builder reportBuilder) {
+
+            // embargo is only applicable to summative exams
+            if (!asmtTypeCode.equalsIgnoreCase(AssessmentType.SUMMATIVE.code())) return rowToVerify;
+
+            //embargo is only applicable to the current year exams
+            if (rowToVerify.getAssessment().getExamSchoolYear() != currentSchoolYear) return rowToVerify;
+
+            final Organization organization = rowToVerify.getOrganization();
```
[the remark this answers is not in the document]
Comment (agorina) [reply aggregate-service/src/main/java/org/opentestsystem/rdw/olap/repository/impl/JdbcCustomAggregateReportRowRepository.java]: We could fast-fail even faster - as soon as at least once `createdWhileDataEmbargoed` is set to true. I can look into making this change. Thank you.

File: common/src/main/java/org/opentestsystem/rdw/reporting/common/model/AggregateResponseMessage.java (hunk 14 above)
Comment (agorina) [reply common/src/main/java/org/opentestsystem/rdw/reporting/common/model/AggregateResponseMessage.java]: That is correct.

**Pull request merged**